File: kore-ruby/kore_fileformat.rb
# frozen_string_literal: true

require 'fiddle'
require 'fiddle/import'

module Kore
  # KORE columnar format bindings for Ruby via Fiddle.
  #
  # Features:
  #   - Read/write KORE v2 binary files
  #   - All 11 ACID features
  #   - Native Ruby types (Arrays, Hashes)
  #   - Idiomatic Ruby API
  #
  # Example:
  #   require 'kore'
  #
  #   # Create data block
  #   block = Kore::DataBlock.new
  #   block.add_column('numbers', Kore::DataType::I64, [1, 2, 3, 4, 5])
  #   block.add_column('names', Kore::DataType::STR, %w[a b c d e])
  #
  #   # Write file
  #   Kore.write_file('/tmp/data.kore', block)
  #
  #   # Read file
  #   restored = Kore.read_file('/tmp/data.kore')
  #   puts "Rows: #{restored.num_rows}, Cols: #{restored.num_columns}"
  class FileFormat
    # ─────────────────────────────────────────────────────────────────────────
    # DATA TYPES & ENUMS
    # ─────────────────────────────────────────────────────────────────────────

    # KORE column data types (must match Rust DType enum).
    module DataType
      I64      = 1  # 64-bit signed integer
      F64      = 2  # 64-bit floating point
      BOOL     = 3  # Boolean
      STR      = 4  # UTF-8 string
      STR_DICT = 5  # Dictionary-encoded string
      ARRAY    = 6  # Nested array
      STRUCT   = 7  # Nested struct
    end

    # KORE compression codecs (must match Rust Compression enum).
    module Compression
      RAW     = 0  # No compression
      RLE     = 1  # Run-length encoding
      DELTA   = 2  # Delta encoding
      DICT    = 3  # Dictionary encoding
      NAN_RAW = 4  # Special NaN handling
      DEFLATE = 5  # Deflate/LZ4
      ZSTD    = 6  # ZSTD compression
    end

    # ─────────────────────────────────────────────────────────────────────────
    # CORE CLASSES
    # ─────────────────────────────────────────────────────────────────────────

    # Column statistics for predicate pushdown.
    class ColumnStats
      attr_accessor :min_value, :max_value, :null_count, :cardinality, :crc32

      def initialize(min_value: nil, max_value: nil, null_count: 0, cardinality: 0, crc32: 0)
        @min_value = min_value
        @max_value = max_value
        @null_count = null_count
        @cardinality = cardinality
        @crc32 = crc32
      end

      def to_h
        {
          min_value: @min_value,
          max_value: @max_value,
          null_count: @null_count,
          cardinality: @cardinality,
          crc32: @crc32,
        }
      end
    end

    # Single column in a data block.
    class Column
      attr_accessor :name, :dtype, :data, :stats

      def initialize(name, dtype, data, stats = nil)
        @name = name
        @dtype = dtype
        @data = data
        @stats = stats
      end

      def to_h
        {
          name: @name,
          type: @dtype,
          data: @data,
          stats: @stats&.to_h,
        }
      end
    end

    # Multi-column data structure.
    class DataBlock
      attr_reader :columns, :num_rows

      def initialize
        @columns = []
        @num_rows = 0
      end

      # Add a column to the data block.
      def add_column(name, dtype, data)
        if @num_rows.zero?
          @num_rows = data.length
        elsif data.length != @num_rows
          raise ArgumentError, "Column '#{name}' has #{data.length} rows, expected #{@num_rows}"
        end

        @columns << Column.new(name, dtype, data)
      end

      # Get column by name.
      def get_column(name)
        @columns.find { |c| c.name == name }
      end

      # Get number of columns.
      def num_columns
        @columns.length
      end

      def to_h
        {
          num_rows: @num_rows,
          num_columns: num_columns,
          columns: @columns.map(&:to_h),
        }
      end
    end

    # MVCC version tracking for time travel.
    class VersionSnapshot
      attr_accessor :version_id, :timestamp, :block_offset, :row_count, :prev_version

      def initialize(version_id:, timestamp:, block_offset:, row_count:, prev_version: nil)
        @version_id = version_id
        @timestamp = timestamp
        @block_offset = block_offset
        @row_count = row_count
        @prev_version = prev_version
      end

      def to_h
        {
          version_id: @version_id,
          timestamp: @timestamp,
          block_offset: @block_offset,
          row_count: @row_count,
          prev_version: @prev_version,
        }
      end
    end

    # Partition evolution support.
    class PartitionSpec
      attr_accessor :spec_id, :columns, :transforms, :parent_spec_id

      def initialize(spec_id:, columns: [], transforms: [], parent_spec_id: nil)
        @spec_id = spec_id
        @columns = columns
        @transforms = transforms
        @parent_spec_id = parent_spec_id
      end

      def to_h
        {
          spec_id: @spec_id,
          columns: @columns,
          transforms: @transforms,
          parent_spec_id: @parent_spec_id,
        }
      end
    end

    # Row-level delete bitmap for soft deletes.
    class DeleteVector
      attr_accessor :bitmap, :cardinality, :timestamp

      def initialize(bitmap: '', cardinality: 0, timestamp: 0)
        @bitmap = bitmap
        @cardinality = cardinality
        @timestamp = timestamp
      end

      def to_h
        {
          bitmap: @bitmap,
          cardinality: @cardinality,
          timestamp: @timestamp,
        }
      end
    end

    # ─────────────────────────────────────────────────────────────────────────
    # ─────────────────────────────────────────────────────────────────────────
    # FFI BINDINGS — Fiddle-based, loads kore_ffi.dll at runtime
    # ─────────────────────────────────────────────────────────────────────────

    @_lib = nil

    class << self
      def load_library
        return @_lib if @_lib

        here = File.dirname(__FILE__)
        candidates = [
          File.join(here, '..', 'target', 'release', 'kore_ffi.dll'),    # Windows
          File.join(here, '..', 'target', 'release', 'libkore_ffi.so'),  # Linux
          File.join(here, '..', 'target', 'release', 'libkore_ffi.dylib'), # macOS
          'libkore_ffi.so', 'libkore_ffi.dylib', 'kore_ffi.dll',
        ]
        path = candidates.find { |c| File.exist?(c) }
        raise 'kore_ffi not found. Build: cargo build --release -p kore-ffi' unless path
        @_lib = Fiddle.dlopen(path)
      end

      def ffi_fn(name, ret, args)
        lib = load_library
        Fiddle::Function.new(lib[name.to_s], args, ret)
      end

      # ─────────────────────────────────────────────────────────────────────
      # HIGH-LEVEL API
      # ─────────────────────────────────────────────────────────────────────

      def crc32(data)
        fn = ffi_fn(:kore_crc32, Fiddle::TYPE_INT, [Fiddle::TYPE_VOIDP, Fiddle::TYPE_SIZE_T])
        fn.call(data, data.bytesize) & 0xFFFF_FFFF
      end

      def write_file(path, data_block)
        block_new  = ffi_fn(:kore_block_new,     Fiddle::TYPE_VOIDP, [])
        block_free = ffi_fn(:kore_block_free,    Fiddle::TYPE_VOID,  [Fiddle::TYPE_VOIDP])
        add_f64    = ffi_fn(:kore_block_add_f64, Fiddle::TYPE_INT,   [Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_SIZE_T])
        add_i64    = ffi_fn(:kore_block_add_i64, Fiddle::TYPE_INT,   [Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_SIZE_T])
        write_fn   = ffi_fn(:kore_write_file,    Fiddle::TYPE_INT,   [Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP])

        handle = block_new.call
        begin
          data_block.columns.each do |col|
            name_ptr = Fiddle::Pointer[col.name + "\0"]
            case col.dtype
            when DataType::F64
              buf = col.data.map(&:to_f).pack('d*')
              add_f64.call(handle, name_ptr, Fiddle::Pointer[buf], col.data.length)
            when DataType::I64
              buf = col.data.map(&:to_i).pack('q*')
              add_i64.call(handle, name_ptr, Fiddle::Pointer[buf], col.data.length)
            end
          end
          rc = write_fn.call(Fiddle::Pointer[path + "\0"], handle)
          raise "kore_write_file failed (rc=#{rc})" unless rc == 0
        ensure
          block_free.call(handle)
        end
      end

      def read_file(path)
        read_fn   = ffi_fn(:kore_read_file,      Fiddle::TYPE_VOIDP, [Fiddle::TYPE_VOIDP])
        block_free= ffi_fn(:kore_block_free,     Fiddle::TYPE_VOID,  [Fiddle::TYPE_VOIDP])
        num_rows  = ffi_fn(:kore_block_num_rows, Fiddle::TYPE_LONG,  [Fiddle::TYPE_VOIDP])
        num_cols  = ffi_fn(:kore_block_num_cols, Fiddle::TYPE_INT,   [Fiddle::TYPE_VOIDP])
        col_name  = ffi_fn(:kore_block_col_name, Fiddle::TYPE_VOIDP, [Fiddle::TYPE_VOIDP, Fiddle::TYPE_SIZE_T])
        get_f64   = ffi_fn(:kore_block_get_f64,  Fiddle::TYPE_LONG,  [Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, -Fiddle::TYPE_LONG_LONG])
        free_str  = ffi_fn(:kore_free_string,    Fiddle::TYPE_VOID,  [Fiddle::TYPE_VOIDP])

        handle = read_fn.call(Fiddle::Pointer[path + "\0"])
        raise "kore_read_file failed: #{path}" if handle.null?
        begin
          nrows = num_rows.call(handle)
          ncols = num_cols.call(handle)
          block = DataBlock.new
          block.instance_variable_set(:@num_rows, nrows)

          ncols.times do |ci|
            name_ptr = col_name.call(handle, ci)
            cname = name_ptr.null? ? "col#{ci}" : name_ptr.to_s
            free_str.call(name_ptr) unless name_ptr.null?

            buf_ptr = Fiddle::Pointer.malloc(nrows * 8)
            n = get_f64.call(handle, Fiddle::Pointer[cname + "\0"], buf_ptr, nrows)
            data = buf_ptr.to_str(n * 8).unpack("d#{n}") if n > 0
            block.add_column(cname, DataType::F64, data || []) if data
          end
          block
        ensure
          block_free.call(handle)
        end
      end

      def read_at_version(data, timestamp)
        raise NotImplementedError, 'Phase 3: Time travel API pending'
      end

      def encrypt_aes256(password, data)
        raise NotImplementedError, 'Phase 3: Encryption API pending'
      end

      def decrypt_aes256(password, encrypted_data)
        raise NotImplementedError, 'Phase 3: Decryption API pending'
      end

      def get_column_stats(data, column_name)
        raise NotImplementedError, 'Phase 3: Stats API pending'
      end

      # Get Bloom filter for a column.
      def get_bloom_filter(data, column_name)
        raise NotImplementedError, 'Phase 3: Bloom filter API pending'
      end
    end

    # ─────────────────────────────────────────────────────────────────────────
    # VERSION
    # ─────────────────────────────────────────────────────────────────────────

    VERSION = '2.0.0'
  end

  # Convenience module-level API
  module_function

  def write_file(path, data)
    FileFormat.write_file(path, data)
  end

  def read_file(path)
    FileFormat.read_file(path)
  end

  def crc32(data)
    FileFormat.crc32(data)
  end

  def encrypt_aes256(password, data)
    FileFormat.encrypt_aes256(password, data)
  end

  def decrypt_aes256(password, data)
    FileFormat.decrypt_aes256(password, data)
  end
end

File: kore-php/KoreFileFormat.php
<?php

/**
 * KORE File Format PHP Extension
 * ==============================
 *
 * PHP bindings to the KORE columnar format via ext-ffi.
 *
 * Features:
 *   - Read/write KORE v2 binary files
 *   - All 11 ACID features
 *   - Native PHP arrays and objects
 *   - Type-safe API
 *
 * Example:
 *   $kore = new Kore\FileFormat();
 *   $block = new Kore\DataBlock();
 *   $block->addColumn('numbers', Kore\DataType::I64, [1, 2, 3, 4, 5]);
 *   $block->addColumn('names', Kore\DataType::STR, ['a', 'b', 'c', 'd', 'e']);
 *   $kore->writeFile('/tmp/data.kore', $block);
 *
 * Requirements:
 *   - PHP 7.4+ (for typed properties)
 *   - PHP FFI extension enabled (php.ini: ffi.enable = "preloaded")
 */

namespace Kore;

// ─────────────────────────────────────────────────────────────────────────────
// DATA TYPES & ENUMS
// ─────────────────────────────────────────────────────────────────────────────

/**
 * KORE column data types (must match Rust DType enum).
 */
abstract class DataType
{
    public const I64 = 1;          // 64-bit signed integer
    public const F64 = 2;          // 64-bit floating point
    public const BOOL = 3;         // Boolean
    public const STR = 4;          // UTF-8 string
    public const STR_DICT = 5;     // Dictionary-encoded string
    public const ARRAY = 6;        // Nested array
    public const STRUCT = 7;       // Nested struct
}

/**
 * KORE compression codecs (must match Rust Compression enum).
 */
abstract class Compression
{
    public const RAW = 0;          // No compression
    public const RLE = 1;          // Run-length encoding
    public const DELTA = 2;        // Delta encoding
    public const DICT = 3;         // Dictionary encoding
    public const NAN_RAW = 4;      // Special NaN handling
    public const DEFLATE = 5;      // Deflate/LZ4
    public const ZSTD = 6;         // ZSTD compression
}

// ─────────────────────────────────────────────────────────────────────────────
// CORE CLASSES
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Column statistics for predicate pushdown.
 */
class ColumnStats
{
    public ?int $minValue = null;
    public ?int $maxValue = null;
    public ?float $minValueF = null;
    public ?float $maxValueF = null;
    public int $nullCount = 0;
    public int $cardinality = 0;
    public int $crc32 = 0;

    public function __construct(
        ?int $minValue = null,
        ?int $maxValue = null,
        ?float $minValueF = null,
        ?float $maxValueF = null,
        int $nullCount = 0,
        int $cardinality = 0,
        int $crc32 = 0
    ) {
        $this->minValue = $minValue;
        $this->maxValue = $maxValue;
        $this->minValueF = $minValueF;
        $this->maxValueF = $maxValueF;
        $this->nullCount = $nullCount;
        $this->cardinality = $cardinality;
        $this->crc32 = $crc32;
    }

    public function toArray(): array
    {
        return [
            'min_value' => $this->minValue,
            'max_value' => $this->maxValue,
            'min_value_f' => $this->minValueF,
            'max_value_f' => $this->maxValueF,
            'null_count' => $this->nullCount,
            'cardinality' => $this->cardinality,
            'crc32' => $this->crc32,
        ];
    }
}

/**
 * Single column in a data block.
 */
class Column
{
    public string $name;
    public int $dtype;
    public array $data;
    public ?ColumnStats $stats = null;

    public function __construct(string $name, int $dtype, array $data, ?ColumnStats $stats = null)
    {
        $this->name = $name;
        $this->dtype = $dtype;
        $this->data = $data;
        $this->stats = $stats;
    }

    public function toArray(): array
    {
        return [
            'name' => $this->name,
            'type' => $this->dtype,
            'data' => $this->data,
            'stats' => $this->stats ? $this->stats->toArray() : null,
        ];
    }
}

/**
 * Multi-column data structure.
 */
class DataBlock
{
    public array $columns = [];
    public int $numRows = 0;

    /**
     * Add a column to the data block.
     */
    public function addColumn(string $name, int $dtype, array $data): void
    {
        if ($this->numRows === 0) {
            $this->numRows = count($data);
        } elseif (count($data) !== $this->numRows) {
            throw new \InvalidArgumentException(
                "Column '{$name}' has " . count($data) . " rows, expected {$this->numRows}"
            );
        }

        $this->columns[] = new Column($name, $dtype, $data);
    }

    /**
     * Get column by name.
     */
    public function getColumn(string $name): ?Column
    {
        foreach ($this->columns as $col) {
            if ($col->name === $name) {
                return $col;
            }
        }
        return null;
    }

    /**
     * Get number of columns.
     */
    public function getNumColumns(): int
    {
        return count($this->columns);
    }

    public function toArray(): array
    {
        return [
            'num_rows' => $this->numRows,
            'num_columns' => $this->getNumColumns(),
            'columns' => array_map(fn ($col) => $col->toArray(), $this->columns),
        ];
    }
}

/**
 * MVCC version tracking for time travel.
 */
class VersionSnapshot
{
    public int $versionId;
    public int $timestamp;
    public int $blockOffset;
    public int $rowCount;
    public ?int $prevVersion = null;

    public function __construct(
        int $versionId,
        int $timestamp,
        int $blockOffset,
        int $rowCount,
        ?int $prevVersion = null
    ) {
        $this->versionId = $versionId;
        $this->timestamp = $timestamp;
        $this->blockOffset = $blockOffset;
        $this->rowCount = $rowCount;
        $this->prevVersion = $prevVersion;
    }

    public function toArray(): array
    {
        return [
            'version_id' => $this->versionId,
            'timestamp' => $this->timestamp,
            'block_offset' => $this->blockOffset,
            'row_count' => $this->rowCount,
            'prev_version' => $this->prevVersion,
        ];
    }
}

/**
 * Partition evolution support.
 */
class PartitionSpec
{
    public int $specId;
    public array $columns = [];
    public array $transforms = [];
    public ?int $parentSpecId = null;

    public function __construct(
        int $specId,
        array $columns = [],
        array $transforms = [],
        ?int $parentSpecId = null
    ) {
        $this->specId = $specId;
        $this->columns = $columns;
        $this->transforms = $transforms;
        $this->parentSpecId = $parentSpecId;
    }

    public function toArray(): array
    {
        return [
            'spec_id' => $this->specId,
            'columns' => $this->columns,
            'transforms' => $this->transforms,
            'parent_spec_id' => $this->parentSpecId,
        ];
    }
}

/**
 * Row-level delete bitmap for soft deletes.
 */
class DeleteVector
{
    public string $bitmap;
    public int $cardinality;
    public int $timestamp;

    public function __construct(string $bitmap = '', int $cardinality = 0, int $timestamp = 0)
    {
        $this->bitmap = $bitmap;
        $this->cardinality = $cardinality;
        $this->timestamp = $timestamp;
    }

    public function toArray(): array
    {
        return [
            'bitmap' => bin2hex($this->bitmap),
            'cardinality' => $this->cardinality,
            'timestamp' => $this->timestamp,
        ];
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// FFI BINDINGS
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Main KORE FileFormat API — wired to Rust kore_ffi via PHP FFI.
 */
class FileFormat
{
    private static ?\FFI $ffi = null;

    private static function getFfi(): \FFI
    {
        if (self::$ffi !== null) {
            return self::$ffi;
        }

        $here = __DIR__;
        $candidates = [
            $here . '/../target/release/kore_ffi.dll',      // Windows
            $here . '/../target/release/libkore_ffi.so',    // Linux
            $here . '/../target/release/libkore_ffi.dylib', // macOS
        ];
        $dll = null;
        foreach ($candidates as $c) {
            if (file_exists($c)) { $dll = realpath($c); break; }
        }
        if (!$dll) {
            throw new \RuntimeException(
                'kore_ffi not found. Build: cargo build --release -p kore-ffi'
            );
        }

        self::$ffi = \FFI::cdef(<<<EOH
            typedef void KoreBlock;
            uint32_t kore_crc32(const uint8_t* data, size_t len);
            KoreBlock* kore_block_new();
            void       kore_block_free(KoreBlock* block);
            int        kore_block_add_f64(KoreBlock* block, const char* name, const double* data, size_t len);
            int        kore_block_add_i64(KoreBlock* block, const char* name, const int64_t* data, size_t len);
            int        kore_write_file(const char* path, KoreBlock* block);
            KoreBlock* kore_read_file(const char* path);
            uint64_t   kore_block_num_rows(const KoreBlock* block);
            uint32_t   kore_block_num_cols(const KoreBlock* block);
            char*      kore_block_col_name(const KoreBlock* block, size_t idx);
            int64_t    kore_block_get_f64(const KoreBlock* block, const char* col, double* out, uint64_t maxlen);
            void       kore_free_string(char* s);
EOH, $dll);
        return self::$ffi;
    }

    // ─────────────────────────────────────────────────────────────────────
    // HIGH-LEVEL API
    // ─────────────────────────────────────────────────────────────────────

    public static function crc32(string $data): int
    {
        $ffi = self::getFfi();
        $buf = \FFI::new("uint8_t[" . strlen($data) . "]");
        \FFI::memcpy($buf, $data, strlen($data));
        return $ffi->kore_crc32($buf, strlen($data));
    }

    public static function writeFile(string $path, DataBlock $dataBlock): void
    {
        $ffi    = self::getFfi();
        $handle = $ffi->kore_block_new();
        if (\FFI::isNull($handle)) throw new \RuntimeException('kore_block_new failed');
        try {
            foreach ($dataBlock->columns as $col) {
                if ($col->dtype === DataType::F64 && count($col->data) > 0) {
                    $arr = \FFI::new('double[' . count($col->data) . ']');
                    foreach ($col->data as $i => $v) $arr[$i] = (float)$v;
                    $ffi->kore_block_add_f64($handle, $col->name, $arr, count($col->data));
                } elseif ($col->dtype === DataType::I64 && count($col->data) > 0) {
                    $arr = \FFI::new('int64_t[' . count($col->data) . ']');
                    foreach ($col->data as $i => $v) $arr[$i] = (int)$v;
                    $ffi->kore_block_add_i64($handle, $col->name, $arr, count($col->data));
                }
            }
            $rc = $ffi->kore_write_file($path, $handle);
            if ($rc !== 0) throw new \RuntimeException("kore_write_file failed (rc=$rc)");
        } finally {
            $ffi->kore_block_free($handle);
        }
    }

    public static function readFile(string $path): DataBlock
    {
        $ffi    = self::getFfi();
        $handle = $ffi->kore_read_file($path);
        if (\FFI::isNull($handle)) throw new \RuntimeException("kore_read_file failed: $path");
        try {
            $nrows = $ffi->kore_block_num_rows($handle);
            $ncols = $ffi->kore_block_num_cols($handle);
            $block = new DataBlock();

            for ($ci = 0; $ci < $ncols; $ci++) {
                $rawName = $ffi->kore_block_col_name($handle, $ci);
                $colName = \FFI::isNull($rawName) ? "col$ci" : \FFI::string($rawName);
                if (!\FFI::isNull($rawName)) $ffi->kore_free_string($rawName);

                $buf = \FFI::new("double[$nrows]");
                $n   = $ffi->kore_block_get_f64($handle, $colName, $buf, $nrows);
                if ($n > 0) {
                    $data = [];
                    for ($i = 0; $i < $n; $i++) $data[] = $buf[$i];
                    $block->addColumn($colName, DataType::F64, $data);
                }
            }
            return $block;
        } finally {
            $ffi->kore_block_free($handle);
        }
    }

    public static function readAtVersion(string $data, int $timestamp): DataBlock
    {
        throw new \RuntimeException('Phase 3: Time travel API pending');
    }

    public static function encryptAes256(string $password, string $data): string
    {
        throw new \RuntimeException('Phase 3: Encryption API pending');
    }

    public static function decryptAes256(string $password, string $encryptedData): string
    {
        throw new \RuntimeException('Phase 3: Decryption API pending');
    }

    /**
     * Get statistics for a column.
     */
    public static function getColumnStats(string $data, string $columnName): ColumnStats
    {
        throw new \RuntimeException('Phase 3: Stats API pending');
    }

    /**
     * Get Bloom filter for a column.
     */
    public static function getBloomFilter(string $data, string $columnName): string
    {
        throw new \RuntimeException('Phase 3: Bloom filter API pending');
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// VERSION
// ─────────────────────────────────────────────────────────────────────────────

const VERSION = '2.0.0';
